crates/mpz-ole-core/src/receiver.rs

```diff
@@ -0,0 +1,166 @@
+    pub fn generate(
+        &self,
+        input: Vec<F>,
+        ot_message_choices: Vec<F>,
+        masked: Vec<MaskedInput<N, F>>,
+    ) -> Result<Vec<ReceiverShare<F>>, OLEError> {
+        if input.len() * F::BIT_SIZE as usize != ot_message_choices.len() {
+            return Err(OLEError::ExpectedMultipleOf(
+                input.len() * F::BIT_SIZE as usize,
+                ot_message_choices.len(),
+            ));
+        }
+
+        if input.len() != masked.len() {
+            return Err(OLEError::WrongNumerOfMasks(masked.len(), input.len()));
+        }
+
+        let shares: Vec<ReceiverShare<F>> = input
+            .iter()
+            .zip(ot_message_choices.chunks_exact(F::BIT_SIZE as usize))
+            .zip(masked)
+            .map(|((&f, chunk), m)| {
+                ReceiverShare::new::<N>(
+                    f,
+                    chunk
+                        .try_into()
+                        .expect("Slice should have length of bit size of field element"),
+                    m,
+                )
+            })
+            .collect();
+
+        Ok(shares)
+    }
+
+    /// Generates new OLEs and stores them internally.
+    ///
+    /// This method is similar to [`OLEReceiver::generate`], except that [`ReceiverShare`]s are stored
+    /// for later adjustment.
+    ///
+    /// # Arguments
+    ///
+    /// * `input` - The receiver's OLE input factors.
+    /// * `ot_message_choices` - The OT messages chosen by the receiver.
+    /// * `masked` - The correlations from the sender.
+    pub fn preprocess(
+        &mut self,
+        input: Vec<F>,
+        ot_message_choices: Vec<F>,
+        masked: Vec<MaskedInput<N, F>>,
+    ) -> Result<(), OLEError> {
+        let shares = self.generate(input, ot_message_choices, masked)?;
+        self.cache.extend(shares);
+
+        Ok(())
+    }
+
+    /// Returns OLEs from internal cache.
+    ///
+    /// For consumption of OLEs which have been stored by [`OLEReceiver::preprocess`].
+    ///
+    /// # Arguments
+    ///
+    /// * `count` - The number of shares to return.
+    ///
+    /// # Returns
+    ///
+    /// * A vector of [`ReceiverShare`]s containing the OLE outputs for the receiver.
+    pub fn consume(&mut self, count: usize) -> Result<Vec<ReceiverShare<F>>, OLEError> {
+        if count > self.cache.len() {
+            return Err(OLEError::InsufficientOLEs(count, self.cache.len()));
+        }
+
+        let shares = self.cache.drain(..count).collect();
+        Ok(shares)
+    }
+
+    /// Adjusts OLEs in the internal cache.
+    ///
+    /// # Arguments
+    ///
+    /// * `targets` - The new OLE receiver inputs.
+    ///
+    /// # Returns
+    ///
+    /// * A vector of [`ReceiverAdjust`] which needs to be converted by [`OLEReceiver::finish_adjust`].
+    /// * A vector of [`ShareAdjust`] which needs to be sent to the [`super::OLESender`].
+    pub fn adjust(
```

Comment:
This would mean that we have to introduce some `HashMap` and ticketing system, right? Because otherwise consumers of this API have to call `adjust` and `finish_adjust` in the right order. So for example if you have consumer `A` and `B` and they do it in this order, then it would **not** work.
`A` -> `adjust`, `B` -> `adjust`, `B` -> `finish_adjust`, `A` -> `finish_adjust`

That's why I chose to return it instead of storing it.